test/test-addon.js

```diff
@@ -26,3 +28,26 @@ test('build simple addon', function (t) {
   proc.stdout.setEncoding('utf-8')
   proc.stderr.setEncoding('utf-8')
 })
+
+test('build with different build dir', function(t){
+  var cmd = [nodeGyp, 'rebuild', '-C', addonPath, '--loglevel=verbose', '--build_dir=foo']
+  var proc = execFile(process.execPath, cmd, function (err, stdout, stderr) {
+    var logLines = stderr.toString().trim().split(/\r?\n/)
+    var lastLine = logLines[logLines.length-1]
+    t.strictEqual(err, null)
+    t.strictEqual(lastLine, 'gyp info ok', 'should end in ok')
+
+    try {
+      var binding = require('hello_world')
+      t.strictEqual(binding.hello(), 'world')
+
+      // Cleanup
+      exec('rm -rf foo', t.end)
```

Comment:
This won't work on Windows.  Can you run `node-gyp --build_dir=foo clean`?

Naming bikeshed: what about calling it `--build-dir`?  That's more in line with `--dist-url`.  (OTOH there's `--msvs_version` but that mimics the gyp variable.)